File: nova_example/bouncing_ball/nova_bouncing_ball.py
from pathlib import Path
import os
import sys

# Add project root to sys.path to import nova_sim_py
project_root = Path(__file__).parent.parent.parent
sys.path.append(str(project_root))

from nova_sim_py import *
from nova_sim_py.plotter import Plotter, TimeSeriesConfig

def main():
    print(f"NovaLib version: {NovaLib.version()}")
    NovaLib.set_log_level("debug")

    # 定位 FMU (路径相对于本文件位置调整)
    fmu_path = str((Path(__file__).parent.parent.parent / 'data' / 'fmus' / '3.0' / 'ref' / 'BouncingBall.fmu').resolve())
    
    result_dir = project_root / 'results' / 'python'
    result_dir.mkdir(parents=True, exist_ok=True)
    result_file = str(result_dir / "nova_bouncing_ball.csv")

    with NovaSimulationStructure() as ss:
        ss.add_model("ball", fmu_path)

        # 1/100s 步长
        with NovaSimulation(structure=ss, step_size=1/100) as sim:
            sim.add_csv_writer(result_file)
            sim.init()
            
            # 仿真 10 秒 (1000步)
            sim.step(1000)
            
            sim.terminate()

    config = TimeSeriesConfig(
        title="BouncingBall",
        y_label="Height[m]",
        identifiers=["ball::h"])
    plotter = Plotter(result_file, config)
    plotter.show() 

    print(f"Nova Python BouncingBall finished. Results: {result_file}")

if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        import traceback
        traceback.print_exc()
        exit(1)

File: nova_example/bouncing_ball/nova_bouncing_ball.cpp
#include "nova/engine/nova_engine.hpp"
#include "nova/components/structure/simulation_structure.hpp"
#include <nova/components/algorithm/fixed_step_algorithm.hpp>
#include <nova/components/recorder/csv_recorder.hpp>
#include <nova/components/logger/logger.hpp>
#include <nova/components/util/chart_plotter.hpp>
#include <filesystem>

using namespace nova_sim;

int main()
{
    set_logging_level(log::level::debug);

    std::filesystem::create_directories(RESULT_FOLDER);
    std::string resultFile{std::string(RESULT_FOLDER) + "/nova_bouncing_ball_cpp.csv"};
    const std::string fmuPath{std::string(DATA_FOLDER) + "/fmus/3.0/ref/BouncingBall.fmu"};

    simulation_structure ss;
    ss.add_model("ball", fmuPath);

    double stepSize = 1.0 / 100.0;
    auto algo = std::make_unique<fixed_step_algorithm>(stepSize);
    
    const auto sim = ss.load(std::move(algo));

    auto csvWriter = std::make_unique<csv_recorder>(resultFile);
    const auto outputPath = csvWriter->output_path();
    csv_config& config = csvWriter->config();
    config.register_variable("ball::h");

    sim->add_listener("csv_recorder", std::move(csvWriter));

    sim->init();
    sim->step(1000); // 10 seconds with 1/100 step size
    sim->terminate();

    log::info("Nova C++ BouncingBall finished.");

    TChartConfig plotConfig;
    plotConfig.addChart(TTimeSeriesChart{
        "bouncing_ball",
        "Time[s]",
        {{{"ball", {{"h"}}}}}});

    plot_csv(outputPath, plotConfig);

    return 0;
}
